### PR TITLE
Split operator tests

### PR DESCRIPTION
**Pull request by pk-g**
Adding Tests for Split operator

### CODE REVIEW

File: onnx/backend/test/case/node/split.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+                output_args = ['output' + str(k) for k in range(len(s) if isinstance(s, list) else s)]
+                node = onnx.helper.make_node(
+                    'Split',
+                    inputs=['input'],
+                    outputs=[arg for arg in output_args],
+                    axis=i,
+                    split=s
+                )
+
+                outputs = np.split(input, np.cumsum(s[:-1]) if isinstance(s,list) else s , i)
```
Comment (houseroad) [inline onnx/backend/test/case/node/split.py]: Similar here, only list

**Comment by houseroad** [comment]
@pk-g friendly ping, could you address the comments, and update the PR?
Thanks.

**Comment by houseroad** [comment]
@pk-g re-ping, could you update the PR?

**Comment by pk-g** [comment]
@houseroad Thanks for the review, addressed the feedback and updated the PR.

**Comment by pk-g** [comment]
@houseroad can you please review the latest changes? thanks

File: onnx/backend/test/case/node/split.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    @staticmethod
+    def export_1d():
+        input = np.array([1., 2., 3., 4., 5., 6.])
+
+        node = onnx.helper.make_node(
+            'Split',
+            inputs=['input'],
+            outputs=['output_1', 'output_2', 'output_3'],
+            axis=0,
+            split=[2]
```
Comment (houseroad) [inline onnx/backend/test/case/node/split.py]: We should not have the split attribute here, because the split length is determined by outputs length.

File: onnx/backend/test/case/node/split.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        input = np.array([[1., 2., 3., 4., 5., 6.],
+                          [7., 8., 9., 10., 11., 12.]])
+
+
+        node = onnx.helper.make_node(
+            'Split',
+            inputs=['input'],
+            outputs=['output_1', 'output_2'],
+            axis=1,
+            split=[3]
```
Comment (houseroad) [inline onnx/backend/test/case/node/split.py]: We should not have the split here if we use equal split.

File: onnx/backend/test/case/node/split.py (hunk 2 above)
Comment (pk-g) [reply onnx/backend/test/case/node/split.py]: Are you suggesting that the following part from defs.cc is not accurate too? 

"If the 'split' argument contains only one element, and that element divides the length of split axis evenly, the tensor is split into equal sized parts with given length."

Shouldn't it be the case that outputs length is determined by 'split' attribute ?

File: onnx/backend/test/case/node/split.py (hunk 3 above)
Comment (pk-g) [reply onnx/backend/test/case/node/split.py]: Are you suggesting that the following part from defs.cc is not accurate too? 

"If the 'split' argument contains only one element, and that element divides the length of split axis evenly, the tensor is split into equal sized parts with given length."

Shouldn't it be the case that outputs length is determined by 'split' attribute ?

File: onnx/backend/test/case/node/split.py (hunk 2 above)
Comment (houseroad) [reply onnx/backend/test/case/node/split.py]: The original doc:
Split a tensor into a list of tensors, along the specified 'axis'. Lengths of the parts can be specified using argument 'split'. Otherwise, the tensor is split to equal sized parts.

If you change the behavior, determine the length according to the split, then it will be a breaking changes. I would like to keep the old behavior. But we need to be more clear on the corner case, i.e., if dim[axis] % length[outputs] != 0.

Comment (pk-g) [reply onnx/backend/test/case/node/split.py]: I am not sure why 'split' attributed should not be used in this case even as per the original doc. 'split' attribute is not an optional attribute in original doc and it's defined as:

(From the original doc)
'split' : list of ints
length of each output

So , one interpretation of this definition is that if 'split' is a list only containing one element, that element becomes length of each output.  My proposed change in definition is intended to add more clarity into this interpretation.

Also, a left-to-right logic may define the flow an op more naturally, i.e.,  it may be more natural to use 'input' characteristics  and 'attribute' (in this case 'split') to generate the 'output',  as compared to using 'output' characteristic to define op behavior or possibily interpolate intended 'attribute'.

Comment (houseroad) [reply onnx/backend/test/case/node/split.py]: https://github.com/onnx/onnx/blob/master/onnx/defs/tensor/defs.cc#L154 shows that split is an optional attribute, also if it is required, you will find 'required' in the description of this attribute.

The reasons I don't want to use split in this case:
1) It is a breaking change, which requires backend changes...
2) if we use split to specify the length, this is redundant with the length of the output (also static information), which can also help us figure the length.

File: onnx/backend/test/case/node/split.py (hunk 3 above)
Comment (linkerzhang) [reply onnx/backend/test/case/node/split.py]: Thank you very much! Lu.

I agree with Lu here. Sorry that I missed this point. "Split" should be either not available or containing same number of elements as number of outputs, and the sum of split elements should be equal to the dimension value along the axis.

**Comment by bddppq** [comment]
@pk-g Any progress on addressing the review comments?

**Comment by pk-g** [comment]
@bddppq : we are gathering more opinion on this op to evaluate if a change of behavior/def is needed.

**Comment by pk-g** [comment]
@houseroad can you please clarify what is exactly preventing us from NOT specifying the split attribute in equal size chunks scenario? i.e., what would be the fundamental difference between equal/not equal size chunks?

**Comment by houseroad** [comment]
@pk-g if you want to equally split the input, you should not have split attribute. Totally depends on the output numbers...

**Comment by houseroad** [comment]
@pk-g could you address my comments? Or you need more information about my suggest?

**Comment by pk-g** [comment]
@houseroad For the purpose of this PR, we can proceed with your suggestion. However, I feel it's worthwhile to get more clarity and opinion on expected behavior of this op from ONNX community. Let's track that design discussion as a separate item and close on this PR. I will update the branch with your suggestion shortly.

**Comment by houseroad** [comment]
@pk-g sure, thanks for the update. :-)

**Comment by pk-g** [comment]
@houseroad Please feel free to take a look at the updated PR that incorporates your proposed changes. Thanks for the review.

**Comment by pk-g** [comment]
@houseroad Friendly Ping, I'd appreciate it if you could review the updated PR.

**Comment by houseroad** [comment]
@pk-g I will do it later today. Thanks for reminding me. :-)

File: onnx/backend/test/case/node/split.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+import onnx
+from ..base import Base
+from . import expect
+
+
+class Split(Base):
+
+    @staticmethod
+    def export_1d():
+        input = np.array([1., 2., 3., 4., 5., 6.])
```
Comment (houseroad) [inline onnx/backend/test/case/node/split.py]: Please explicitly specify the dtype=float here.

File: onnx/backend/test/case/node/split.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+        node = onnx.helper.make_node(
+            'Split',
+            inputs=['input'],
+            outputs=['output_1', 'output_2', 'output_3'],
+            axis=0
+        )
+
+        expected_outputs = [np.array([1, 2]), np.array([3, 4]), np.array([5, 6])]
```
Comment (houseroad) [inline onnx/backend/test/case/node/split.py]: Also dtype here.

File: onnx/backend/test/case/node/split.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+        node = onnx.helper.make_node(
+            'Split',
+            inputs=['input'],
+            outputs=['output_1', 'output_2'],
+            axis=0,
+            split=[2, 4]
+        )
+
+        expected_outputs = [np.array([1., 2.]), np.array([3., 4., 5., 6.])]
```
Comment (houseroad) [inline onnx/backend/test/case/node/split.py]: dtype please

File: onnx/backend/test/case/node/split.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        expect(node, inputs=[input], outputs=[y for y in expected_outputs], name='test_split_variable_parts_1d')
+
+    @staticmethod
+    def export_2d():
+        input = np.array([[1., 2., 3., 4., 5., 6.],
```
Comment (houseroad) [inline onnx/backend/test/case/node/split.py]: dtype please

File: onnx/backend/test/case/node/split.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+                          [7., 8., 9., 10., 11., 12.]])
+
+        node = onnx.helper.make_node(
+            'Split',
+            inputs=['input'],
+            outputs=['output_1', 'output_2'],
+            axis=1
+        )
+
+        expected_outputs = [np.array([[1., 2., 3.], [7., 8., 9.]]),
```
Comment (houseroad) [inline onnx/backend/test/case/node/split.py]: dtype please

File: onnx/backend/test/case/node/split.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+        node = onnx.helper.make_node(
+            'Split',
+            inputs=['input'],
+            outputs=['output_1', 'output_2'],
+            axis=1,
+            split=[2, 4]
+        )
+
+        expected_outputs = [np.array([[1., 2.], [7., 8.]]),
```
Comment (houseroad) [inline onnx/backend/test/case/node/split.py]: dtype please

File: onnx/defs/tensor/defs.cc [hunk 10]
```diff
@@ -153,11 +153,11 @@ ONNX_OPERATOR_SCHEMA(Split)
         "Constrain input types to float tensors.")
     .Attr("axis", "Which axis to split on", AttributeProto::INT, OPTIONAL)
     .Attr("split", "length of each output", AttributeProto::INTS, OPTIONAL)
-    .SetDoc(R"DOC(Split a tensor into a list of tensors, along the specified
+    .SetDoc(R"DOC(Split a tensor into a list of tensors, along the specified 
 'axis'. Lengths of the parts can be specified using argument 'split'.
-Otherwise, the tensor is split to equal sized parts.
-)DOC");
-
+Otherwise, the tensor is split to equal sized parts.
```
[the remark this answers is not in the document]
Comment (pk-g) [reply onnx/defs/tensor/defs.cc]: fixed

File: onnx/defs/tensor/defs.cc [hunk 11]
```diff
@@ -153,11 +153,11 @@ ONNX_OPERATOR_SCHEMA(Split)
...
     .Attr("split", "length of each output", AttributeProto::INTS, OPTIONAL)
-    .SetDoc(R"DOC(Split a tensor into a list of tensors, along the specified
+    .SetDoc(R"DOC(Split a tensor into a list of tensors, along the specified 
 'axis'. Lengths of the parts can be specified using argument 'split'.
-Otherwise, the tensor is split to equal sized parts.
-)DOC");
-
+Otherwise, the tensor is split to equal sized parts. 
+ )DOC"); 
+
```
[the remark this answers is not in the document]
Comment (pk-g) [reply onnx/defs/tensor/defs.cc]: fixed

File: onnx/defs/tensor/defs.cc [hunk 12]
```diff
@@ -153,11 +153,11 @@ ONNX_OPERATOR_SCHEMA(Split)
...
     .Attr("axis", "Which axis to split on", AttributeProto::INT, OPTIONAL)
     .Attr("split", "length of each output", AttributeProto::INTS, OPTIONAL)
-    .SetDoc(R"DOC(Split a tensor into a list of tensors, along the specified
+    .SetDoc(R"DOC(Split a tensor into a list of tensors, along the specified 
 'axis'. Lengths of the parts can be specified using argument 'split'.
-Otherwise, the tensor is split to equal sized parts.
-)DOC");
-
+Otherwise, the tensor is split to equal sized parts. 
+ )DOC");
```
[the remark this answers is not in the document]
Comment (pk-g) [reply onnx/defs/tensor/defs.cc]: fixed

File: onnx/backend/test/case/node/split.py (hunk 4 above)
Comment (pk-g) [reply onnx/backend/test/case/node/split.py]: fixed

File: onnx/backend/test/case/node/split.py (hunk 5 above)
Comment (pk-g) [reply onnx/backend/test/case/node/split.py]: fixed

File: onnx/backend/test/case/node/split.py (hunk 6 above)
Comment (pk-g) [reply onnx/backend/test/case/node/split.py]: fixed

File: onnx/backend/test/case/node/split.py (hunk 7 above)
Comment (pk-g) [reply onnx/backend/test/case/node/split.py]: fixed

File: onnx/backend/test/case/node/split.py (hunk 8 above)
Comment (pk-g) [reply onnx/backend/test/case/node/split.py]: fixed

File: onnx/backend/test/case/node/split.py (hunk 9 above)
Comment (pk-g) [reply onnx/backend/test/case/node/split.py]: fixed

**Comment by pk-g** [comment]
@houseroad  can you elaborate more on what you mean by "add the default value 0 for axis" ? we already have two cases for axis = 0 on export_1d

**Comment by pk-g** [comment]
@houseroad ping ... can you please provide more details on previous question so we can move forward with completing this PR? thanks

**Comment by houseroad** [comment]
I mean, you should some cases which do not have axis attribute. So you can check the behavior when default value is used.

**Comment by pk-g** [comment]
@houseroad

**Comment by pk-g** [comment]
@houseroad please feel free to review the updated PR with additional test cases for default axis.

**Comment by houseroad** [comment]
@pk-g thanks for the update, I will check it soon. :-)

File: onnx/defs/tensor/defs.cc [hunk 13]
```diff
@@ -157,7 +157,7 @@ ONNX_OPERATOR_SCHEMA(Split)
 'axis'. Lengths of the parts can be specified using argument 'split'.
 Otherwise, the tensor is split to equal sized parts.
 )DOC");
-
+
```
Comment (houseroad) [inline onnx/defs/tensor/defs.cc]: Nit one space here

File: onnx/backend/test/case/node/split.py [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+        expected_outputs = [np.array([[1., 2.], [7., 8.]]).astype(np.float32),
+                            np.array([[3., 4., 5., 6.], [9., 10., 11., 12.]]).astype(np.float32)]
+
+        expect(node, inputs=[input], outputs=[y for y in expected_outputs], name='test_split_variable_parts_2d')
+
+    @staticmethod
+    def export_default_values():
+        input = np.array([1., 2., 3., 4., 5., 6.]).astype(np.float32)
+
+        node = onnx.helper.make_node(
```
Comment (houseroad) [inline onnx/backend/test/case/node/split.py]: Nit: add comment, tell users our default axis is 0

**Comment by houseroad** [comment]
I mean change the description of attribute axis and regenerate the doc. Thanks! @pk-g

File: onnx/backend/test/case/node/split.py (hunk 14 above)
Comment (pk-g) [reply onnx/backend/test/case/node/split.py]: Not sure why we need to clarify that? The axis of type INT, and it's clear that INT has a default value of 0.

Comment (houseroad) [reply onnx/backend/test/case/node/split.py]: https://github.com/onnx/onnx/blob/master/docs/Operators.md#attributes-56, in this case, it's 1. We have to explicit say it.

Even in numpy, the doc says default axis is 0. https://docs.scipy.org/doc/numpy/reference/generated/numpy.split.html

File: onnx/defs/tensor/defs.cc (hunk 13 above)
Comment (pk-g) [reply onnx/defs/tensor/defs.cc]: removed

File: onnx/backend/test/case/node/split.py (hunk 14 above)
Comment (pk-g) [reply onnx/backend/test/case/node/split.py]: Sure

**Comment by pk-g** [comment]
@houseroad I've updated the PR with latest suggestion on default axis clarification. Please feel free to review so we can close on this PR. Thanks!

File: onnx/defs/tensor/defs.cc [hunk 15]
```diff
@@ -151,7 +151,7 @@ ONNX_OPERATOR_SCHEMA(Split)
         "T",
         {"tensor(float16)", "tensor(float)", "tensor(double)"},
         "Constrain input types to float tensors.")
-    .Attr("axis", "Which axis to split on", AttributeProto::INT, OPTIONAL)
+    .Attr("axis", "Which axis to split on (default to 0)", AttributeProto::INT, OPTIONAL)
```
Comment (houseroad) [inline onnx/defs/tensor/defs.cc]: OPTIONAL ==> static_cast<int64_t>(0)
https://github.com/onnx/onnx/blob/master/onnx/defs/tensor/defs.cc#L327

Comment (pk-g) [reply onnx/defs/tensor/defs.cc]: Sure, thanks!

**Comment by houseroad** [comment]
@pk-g I will do it soon.
